fby3.5: rf: Update the GPIO configurations.

Scron-Chang opened the pull request:
- Change the configurations of default state, output type, direction:

    - GPIOA2:
      Set GPIOA2 as default low to not enable the recovery mode of the
      CXL controller.

    - GPIOA6:
      Set as output and default high because BIC may need to reset JTAG
      directly in a particular case.

    - GPIOA7:
      Set as default low to set the JTAG mode.

    - GPIOB1:
      Set as default low to set CXL controller boots from its own ROM in
      the default situation.

    - GPIOB3:
      Set as input to not control the CXL controller GPIO_0.

    - GPIOB6:
      Set as default high to avoid resetting the I2C interface of the
      CXL controller when BIC reboots.

    - GPIOF0:
      Set as default high and open-drain output type. This pin controls
      the flash memory which uses 1.8V.

    - GPIOF1:
      Set as default high to match the default HW state.

    - GPIOH2:
      Set as Output and default low to enable the SPI buffer in the
      default mode.

- Enable the latch feature for the following GPIOs to avoid making
  unexpected consequences:

    - GPIOA0/1:
      These two pins trigger the cold and warm reset signal on the CXL
      controller, respectively.

    - GPIOC0~D1:
      These pin control the power deliver.

    - GPIOE1:
      This pin is used to inform the mainboard that Rainbow Falls is
      ready.

    - GPIOF2:
      This pin controls the buffer handling the voltage shift.

Tested:
- Rainbow Falls platfrom build: Pass.
- BIC warm reboot doesn't affect the PM8702. I have verified this from
  the PM8702 console.
- Console logs from the RF BIC:
```
[set_MB_DC_status] gpio number(32) status(1)
[control_power_on_sequence] power on success
[set_DC_status] gpio number(33) status(1)

uart:~$ platform gpio list_all
[0  ] ASIC_DEV_RST_N                     : OD  | output(I) | 1(1)
[1  ] ASIC_PERST0_N                      : OD  | output(I) | 1(1)
[2  ] ASIC_PERST1_N                      : OD  | output(O) | 0(0)
[6  ] JTAG2_BIC_ASIC_NTRST2              : PP  | output(O) | 1(1)
[7  ] ASIC_TAP_SEL                       : OD  | output(O) | 0(0)
[9  ] ASIC_CPU_BOOT_1                    : OD  | output(O) | 0(0)
[11 ] ASIC_GPIO_R_0                      : OD  | input (I) | 1(1)
[14 ] I2CS_SRSTB_GPIO                    : OD  | output(I) | 1(1)
[35 ] SPI_MASTER_SEL                     : PP  | output(O) | 0(0)
[40 ] SPI_RST_FLASH_N                    : OD  | output(I) | 0(0)
[41 ] SMBUS_ALERT_R_N                    : OD  | output(I) | 1(1)
[58 ] SPI_BIC_SHIFT_EN                   : OD  | output(O) | 0(0)
```

Signed-off-by: Scron Chang <Scron.Chang@quantatw.com>

GoldenBug commented:
@Scron-Chang 

Hello, Could you please complete the CLA. It's not possible to continue until that is finished.

Scron-Chang commented:
Hi @GoldenBug,
Sorry for the confusion. My CLA is now under process.

GoldenBug commented on meta-facebook/yv35-rf/src/platform/plat_gpio.c:
```diff
@@ -16,59 +16,47 @@ const char *const gpio_name[] = {
-	  NULL },
-	{ CHIP_GPIO, 2, ENABLE, DISABLE, GPIO_OUTPUT, GPIO_HIGH, OPEN_DRAIN, GPIO_INT_DISABLE,
-	  NULL },
+	{ CHIP_GPIO, 0, ENABLE, ENABLE, GPIO_OUTPUT, GPIO_LOW, PUSH_PULL, GPIO_INT_DISABLE, NULL },
```
cannot be push-pull. It needs to be open-drain output

Scron-Chang replied:
Hi @GoldenBug 
Thank you for your responses.

Considering my testing environment is incomplete (Our RainbowFalls lacks the pm8702). We decided to set the GPIOA0~A2 as the push-pull output type because push-pull is more reliable than open-drain.
In addition, because there are buffers between the AST1030 and the pm8702, we can use the push-pull without considering voltage level differences.

I will test whether setting as the open-drain output type is OK for these three GPIO pins after I get the complete RainbowFalls board. If it is, I will update the GPIO table then.

GoldenBug commented on meta-facebook/yv35-rf/src/platform/plat_gpio.c:
```diff
@@ -16,59 +16,47 @@ const char *const gpio_name[] = {
-	{ CHIP_GPIO, 2, ENABLE, DISABLE, GPIO_OUTPUT, GPIO_HIGH, OPEN_DRAIN, GPIO_INT_DISABLE,
-	  NULL },
+	{ CHIP_GPIO, 0, ENABLE, ENABLE, GPIO_OUTPUT, GPIO_LOW, PUSH_PULL, GPIO_INT_DISABLE, NULL },
+	{ CHIP_GPIO, 1, ENABLE, ENABLE, GPIO_OUTPUT, GPIO_LOW, PUSH_PULL, GPIO_INT_DISABLE, NULL },
+	{ CHIP_GPIO, 2, ENABLE, DISABLE, GPIO_OUTPUT, GPIO_LOW, PUSH_PULL, GPIO_INT_DISABLE, NULL },
```
Both should be an open drain output because this signal is pulled high to 1.8V. 
PP will also work, but OD is better.

Scron-Chang replied:
Done.

GoldenBug commented on meta-facebook/yv35-rf/src/platform/plat_gpio.c:
```diff
@@ -100,10 +88,9 @@ GPIO_CFG plat_gpio_cfg[] = {
 	  NULL },
 	/** Group F: 40-47 **/
 	{ CHIP_GPIO, 40, ENABLE, DISABLE, GPIO_INPUT, GPIO_LOW, PUSH_PULL, GPIO_INT_DISABLE, NULL },
```
This should be open drain output, not changed here but should fix it in this PR anyway

Scron-Chang replied:
OK, I think it is OK to fix it in this PR.

Scron-Chang replied:
Done.

Scron-Chang commented:
I rebased this branch and updated the description of this PR. In addition, I removed the log of the BIC console because the RainbowFalls board I am using is incomplete and could cause confusion to the reviewer. If I get the complete RainbowFalls, I will check this PR with it immediately.

GoldenBug commented:
Any updates on the CLA?

Scron-Chang commented:
> Any updates on the CLA?

My manager has sent a mail for my CLA, but we don't receive any response from Meta. Would you kindly check the email box for that?

Scron-Chang commented:
I received the Rainbow Falls board. I updated the code based on the suggestions from @GoldenBug and tested it. Also, the description of this PR has been updated. There now are the result of the gpio list and the reaction of PM8702 when BIC reboots.

Scron-Chang commented:
Hi @GoldenBug ,
I feel excited I just got the CLA.
I just rebased this PR. Please review it again. Appreciate.
